File: models/Gpt-3.5-turbo/cpp/code/problems/additional_problems/Chess_Tournament.cpp

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
    int n;
    cin >> n;

    vector<pair<int, int>> players; // (number of games, player number)

    for (int i = 1; i <= n; ++i) {
        int x;
        cin >> x;
        players.push_back({x, i});
    }

    sort(players.begin(), players.end());

    vector<pair<int, int>> games;

    for (int i = 0; i < n; ++i) {
        int numGames = players[i].first;
        int player = players[i].second;

        for (int j = 0; j < numGames; ++j) {
            if (i + j + 1 >= n) {
                cout << "IMPOSSIBLE" << endl;
                return 0;
            }

            int opponent = players[i + j + 1].second;
            games.push_back({player, opponent});
        }
    }

    cout << games.size() << endl;
    for (auto& game : games) {
        cout << game.first << " " << game.second << endl;
    }

    return 0;
}

File: models/Gpt-3.5-turbo/py/code/additional_problems/Chess_Tournament.py

n = int(input())
games = list(map(int, input().split()))

players = [(i+1, games[i]) for i in range(n)]
players.sort(key=lambda x: x[1], reverse=True)

pairs = []
while len(players) > 1:
    player1 = players[0]
    player2 = players[-1]
    if player1[1] == 0 or player2[1] == 0:
        break
    pairs.append((player1[0], player2[0]))
    players[0] = (player1[0], player1[1] - 1)
    players[-1] = (player2[0], player2[1] - 1)
    players.sort(key=lambda x: x[1], reverse=True)

if len(pairs) == n // 2:
    print(len(pairs))
    for pair in pairs:
        print(pair[0], pair[1])
else:
    print("IMPOSSIBLE")
